### ISSUE TITLE
Mac builds broken when USE_DISTRIBUTED=0

### ISSUE THREAD

**Issue by aslvrstn** [pytorch/pytorch#86448]
### 🐛 Describe the bug

It looks like a default install from HEAD right now on macs will raise

```
ImportError: Torch.distributed is not available, set         ``USE_DISTRIBUTED=1`` to enable it when building PyTorch from source.
```

on simply `import torch`

This looks caused by https://github.com/pytorch/pytorch/pull/85781 . Checking out at the commit before that and everything is happy, and checking out at that commit, (after `python setup.py develop` in either case) and simply running

`python test/test_mps.py` gives:

```
Traceback (most recent call last):
  File "/Users/alex/git/pytorch/test/test_mps.py", line 14, in <module>
    import torch
  File "/Users/alex/git/pytorch/torch/__init__.py", line 805, in <module>
    from .functional import *  # noqa: F403
  File "/Users/alex/git/pytorch/torch/functional.py", line 8, in <module>
    import torch.nn.functional as F
  File "/Users/alex/git/pytorch/torch/nn/__init__.py", line 1, in <module>
    from .modules import *  # noqa: F403
  File "/Users/alex/git/pytorch/torch/nn/modules/__init__.py", line 2, in <module>
    from .linear import Identity, Linear, Bilinear, LazyLinear
  File "/Users/alex/git/pytorch/torch/nn/modules/linear.py", line 7, in <module>
    from .. import functional as F
  File "/Users/alex/git/pytorch/torch/nn/functional.py", line 18, in <module>
    from .._jit_internal import boolean_dispatch, _overload, BroadcastingList1, BroadcastingList2, BroadcastingList3
  File "/Users/alex/git/pytorch/torch/_jit_internal.py", line 39, in <module>
    import torch.distributed.rpc
  File "/Users/alex/git/pytorch/torch/distributed/__init__.py", line 21, in <module>
    raise ImportError(
ImportError: Torch.distributed is not available, set         ``USE_DISTRIBUTED=1`` to enable it when building PyTorch from source.
```

As far as I can tell, what's happening is that torch/nn/functional.py has a transitive dependency on that module, simply in order to be able to call `is_available()`, (here: https://github.com/pytorch/pytorch/blob/master/torch/_jit_internal.py#L1040) but the breaking change caused it to be impossible to load that module at all.

My main confusion is how CI is passing at all for mac / MPS runs. My hunch is that somehow USE_DISTRIBUTED=1 is getting set even for mac builds on CI, despite the default being false on macs (allegedly!)

If someone who knows the build system, particularly the setup for mac CI builds, could weigh in, that would be much appreciated.

Thanks, all!

### Versions

(pytorchdev) alex@Alexs-MacBook-Pro-2 pytorch % python torch/utils/collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.201)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.14.0a0+git9ceadca
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] numpy-base                1.23.1          py310h742c864_0
[conda] torch                     1.14.0a0+git9ceadca           dev_0    <develop>

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @malfet @albanD

**Comment by aslvrstn** [comment]
Confirmed that a clean rebuild with `USE_DISTRIBUTED=1 python setup.py develop` causes things to work again, so further evidence for my hunch.

**Comment by aslvrstn** [comment]
Suspicious looking USE_DISTRIBUTED=1 in a mac CI build: https://github.com/pytorch/pytorch/actions/runs/3201392240/jobs/5229317745#step:9:488

So I'm starting to suspect that the other half of this bug is that nothing in the CI pipeline tests USE_DISTRIBUTED=0

**Comment by aslvrstn** [comment]
Yup, USE_DISTRIBUTED=1 set here: https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/macos-build.sh#L38

and called by https://github.com/pytorch/pytorch/blob/master/.github/workflows/_mac-build.yml#L125

Going to attempt a PR to remove that env var and see if I can reproduce this on CI

**Comment by malfet** [comment]
Thank you for the report. Change indeed looks a bit suspicious, let's see if I can forward fix it quickly

**Comment by aslvrstn** [comment]
Ooh, cool, confirmed that I can reproduce this on CI: https://app.circleci.com/pipelines/github/pytorch/pytorch/578568/workflows/ef0b86ba-b300-4c22-831f-a6433c031fd0/jobs/17159488 failed with
```
ImportError: Torch.distributed is not available, set         ``USE_DISTRIBUTED=1`` to enable it when building PyTorch from source.
```


One confusing thing is that the ios builds, not even the tests, seem to have failed somewhat silently:

```
Oct 07 06:35:06 [ 88%] Built target all_microkernels
Oct 07 06:35:06 make: *** [all] Error 2

Exited with code exit status 2
CircleCI received exit code 2
```

Not really sure what's going on there.


Would be nice to be able to change the mac CI build to align with the default build to make sure this path is exercised

**Comment by malfet** [comment]
Ok, continuing along those lines can help, but it is downright ugly:
```
diff --git a/torch/_jit_internal.py b/torch/_jit_internal.py
index 223d96aa84..c9faacaad2 100644
--- a/torch/_jit_internal.py
+++ b/torch/_jit_internal.py
@@ -36,7 +36,13 @@ import torch
 # This is needed. `torch._jit_internal` is imported before `torch.distributed.__init__`.
 # Explicitly ask to import `torch.distributed.__init__` first.
 # Otherwise, "AttributeError: module 'torch' has no attribute 'distributed'" is raised.
-import torch.distributed.rpc
+try:
+    import torch.distributed.rpc
+    _DISTRIBUTED_RPC_IS_AVAILABLE = torch.distributed.rpc.is_available()
+except ImportError:
+    # This can happen if PyTorch is compiled with USE_DISTRIBUTED=0
+    _DISTRIBUTED_RPC_IS_AVAILABLE = hasattr(torch._C, "_c10d_init")
+    pass
 import torch.package._mangling as package_mangling
 from torch._C import Future as CFuture
 from torch._sources import fake_range, get_source_lines_and_file, parse_def
@@ -1037,7 +1043,7 @@ def is_future(ann) -> bool:
     return getattr(ann, "__origin__", None) is Future
 
 
-if torch.distributed.rpc.is_available():
+if _DISTRIBUTED_RPC_IS_AVAILABLE:
     from torch._C._distributed_rpc import PyRRef
     from torch.distributed.rpc import RRef
 
diff --git a/torch/nn/modules/_functions.py b/torch/nn/modules/_functions.py
index 66200345cb..30950629fe 100644
--- a/torch/nn/modules/_functions.py
+++ b/torch/nn/modules/_functions.py
@@ -1,5 +1,9 @@
 import torch
-import torch.distributed as dist
+try:
+    import torch.distributed as dist
+except ImportError:
+    # This can happen if PyTorch is compiled without distributed support
+    dist = None
 
 from torch.autograd.function import Function
 
diff --git a/torch/nn/parallel/distributed.py b/torch/nn/parallel/distributed.py
index 2e271f7bb0..bb7f5d3698 100644
--- a/torch/nn/parallel/distributed.py
+++ b/torch/nn/parallel/distributed.py
@@ -11,27 +11,34 @@ import warnings
 from contextlib import contextmanager
 
 import torch
-import torch.distributed as dist
+try:
+    import torch.distributed as dist
+    from torch.distributed.algorithms.join import (
+        Join,
+        Joinable,
+        JoinHook,
+    )
+    RPC_AVAILABLE = False
+    if dist.is_available():
+        from torch.distributed.utils import (
+            _verify_param_shape_across_processes,
+            _sync_module_states,
+            _to_kwargs,
+        )
+        from torch.distributed.distributed_c10d import ReduceOp, _get_default_group
+    if torch.distributed.rpc.is_available():
+        RPC_AVAILABLE = True
+        from torch.distributed.rpc import RRef
+except ImportError:
+    RPC_AVAILABLE = False
+    class JoinHook:
+        pass
+    class Joinable:
+        pass
 from torch.autograd import Function, Variable
-from torch.distributed.algorithms.join import (
-    Join,
-    Joinable,
-    JoinHook,
-)
 
 from torch.utils._pytree import tree_flatten, tree_unflatten
 
-RPC_AVAILABLE = False
-if dist.is_available():
-    from torch.distributed.utils import (
-        _verify_param_shape_across_processes,
-        _sync_module_states,
-        _to_kwargs,
-    )
-    from torch.distributed.distributed_c10d import ReduceOp, _get_default_group
-if torch.distributed.rpc.is_available():
-    RPC_AVAILABLE = True
-    from torch.distributed.rpc import RRef
 
 from torch._utils import _get_device_index

```
I'm going to revert the change

**Comment by aslvrstn** [comment]
Confirmed locally that the revert fixes this.

I rebased the PR to change the CI build and it's running now. Curious to see if it runs cleanly: https://github.com/pytorch/pytorch/pull/86451

**Comment by albanD** [comment]
How was this not caught on CI? We build with distributed disabled for our CI as well right?

**Comment by aslvrstn** [comment]
As far as I can tell, no. See my comments from about 10 hours ago. 

> Yup, USE_DISTRIBUTED=1 set here: https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/macos-build.sh#L38
> and called by https://github.com/pytorch/pytorch/blob/master/.github/workflows/_mac-build.yml#L125
> Going to attempt a PR to remove that env var and see if I can reproduce this on CI

**Comment by kulinseth** [comment]
Do we need to similarly [platform check](https://github.com/pytorch/pytorch/pull/85781/files#diff-6464afc7c5c346ed36236ec8e8b052a5db9f786fbccffc9aaa2db665a66c646cR53) for macOS ?

**Comment by albanD** [comment]
I don't think that raising an error on import is an ok thing to do.
1) You won't be able to have `torch.distributed` anymore. The user will have to do `from torch import distributed` to access it.
2) It will keep breaking the public API checking. The API should remain consistent whatever C compilation flags (similar to the rest of the lib).

**Comment by aslvrstn** [comment]
Should be fixed as of #86451 being merged
